pkg/sql/plan2/explain/build_query_explain.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package explain
+
+import (
+	"fmt"
+	// "container/list"
+	// "fmt"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+)
+
+var _ ExplainQuery = &ExplainQueryImpl{}
+
+type ExplainQueryImpl struct {
+	QueryPlan *plan.Query
+}
+
+func NewExplainQueryImpl(query *plan.Query) *ExplainQueryImpl {
+	return &ExplainQueryImpl{
+		QueryPlan: query,
+	}
+}
+
+func (e *ExplainQueryImpl) ExplainPlan(buffer *ExplainDataBuffer, options *ExplainOptions) {
+	var Nodes []*plan.Node = e.QueryPlan.Nodes
+	for index, rootNodeId := range e.QueryPlan.Steps {
+		fmt.Printf("-----------------------------Printing the %v-th plan tree -----------------------\n", index)
+		settings := FormatSettings{
+			buffer:      buffer,
+			offset:      0,
+			indent:      4,
+			indent_char: ' ',
+			level:       0,
+		}
+
+		stack := NewStack()
+		stack.Push(&Frame{
```

Comment:
It is a level traverse. It really needs the stack ?

Reply:
Actually, it is a dfs but not the level traverse.

Reply:
I use depth first search to traverse the query，make sure that the tree structure can be displayed normally

Reply:
DFS walk should not need an explicit stack.   your recursive call stack is your node stack.

Reply:
Yes, I have traversed the plan tree recursively